### ISSUE TITLE
Allow mixing include and exclude patterns for backup and restore

### ISSUE THREAD

**Issue by fd0** [restic/restic#233]
This is a followup issue for #226. Up to now it is only possible to specify exclude patterns for backup and restore.

There are uses cases that aren't covered, for example: A user would like to backup her home directory at `/home/user`, excluding anything from the directory `work` except C source files.

Are there any other uses I haven't thought of?

Implementation/User Interface: Allow specifying both `--include` and `--exclude` patterns on the command line, which fill a common list (order does matter here). For each dir/file, check all patterns in the list. The action (exclude or include) of the last mattching pattern wins. The default action (which will also be used if neither include nor exclude patterns are specified) is "include".

This already contains a corner case: Should restic walk `/home/user/work` at all?

I think it should not. Instead, if this is a desirable behaviour, require the user to add a more specific pattern to signal that the excluded directory should also be walked, e.g. `restic backup --exclude /home/user/work --include /home/user/work/**/*c. /home/user`.

A better user interface would be to allow specifying a file to read the patterns from (`--pattern-file` or something like that). In this file, all lines starting with `#` are comments, empty lines ignored, all other lines must start with either `+` (include) or `-` (exclude) following a space character and a pattern. For the use case above a filter file would look like this:

```
# filter out everything from work, but include c source code files
- /home/user/work
+ /home/user/work/**/*.c
```

This issue can be closed once a definitive solution for having both include and exclude patterns is implemented.

**Comment by fw42** [comment]
Another possibility would be to read an exact list of files to backup from a file (or from stdin). Then people could just use whatever tools they want (find, grep, etc.) to build their list and "pipe" that list to restic.

**Comment by fw42** [comment]
I should read the whole text before posting a reply. Sorry.

**Comment by fd0** [comment]
This is a different issue, taking the list of files/dirs to backup from stdin instead of command line arguments. Do you think it's valuable to have that? If so, could you add an issue?

**Comment by fd0** [comment]
Just a reminder for myself: This issue is about include filters for backup, restore already has them.

**Comment by fd0** [comment]
Just a reminder for myself: This issue is about include filters for backup, restore already has them.

**Comment by lathspell** [comment]
You should maybe just copy rsync's --exclude/--include/--exclude-from syntax, they have 20 years of experience :-)
(at least please add some examples of the current syntax to the user guide as it's unclear if "/foo" and "foo" is thesame or if "*.c" is supported.

**Comment by fd0** [comment]
Ah, thanks for the comment, I've added an issue for the missing examples in the manual: https://github.com/restic/restic/issues/396

To be honest, I don't like the filter syntax from rsync at all, because the rules are way too complex. But we'll see what we can come up with.

**Comment by opennota** [comment]
Any updates? It is definitely a must-have option. Currently it is not even possible to read the file list to back up from the stdin:

```
$ find -name '*.go' | ./restic backup --files-from -
open -: no such file or directory
```

**Comment by fd0** [comment]
Uhm, reading the list of files from stdin can be achieved by calling restic as follows:
```
$ find -name '*.go' | restic backup --files-from /dev/stdin
```

If you like, I'd accept a PR that adds handling `-` for `--files-from`. :)

**Comment by fd0** [comment]
@opennota would you mind describing your use case? It'd be interesting for us.

**Comment by fd0** [comment]
The dash (`-`) issue is tracked as #769.

**Comment by opennota** [comment]
@fd0 
Not much of a use case. I just want to back up only the files with certain extensions without using a temporary file for the list.

> If you like, I'd accept a PR that adds handling - for --files-from. :)

I'll see what I can do.

**Comment by runfalk** [comment]
You can kind of emulate this behavior using ``sed`` and named pipes:

```bash
restic --exclude-file <(sed -n 's/^- \(.*\)/\1/p' files.list) --files-from <(sed -n 's/^+ \(.*\)/\1/p' files.list)
```

Lines starting with ``-`` are excluded and those with ``+`` are included.

**Comment by dgasaway** [comment]
I think a good model is what borg recently implemented for --pattern and --patterns-from
https://borgbackup.readthedocs.io/en/stable/usage/help.html#borg-help-patterns
Not so much the different style selectors, but the options to specify root paths, include rules, exclude rules, and no-recurse exclude rules in a file.

**Comment by blacksails** [comment]
Why not just use the standard ignore format that `.gitignore` uses? eg.
```
# ingore everything
*
# include $HOME/.local
!$HOME/.local
```

**Comment by tnatanael** [comment]
The --include is scheduled to be implemented?

Sound like --include and --exclude cannot be implemented together, or at least would be a hierarchy of precedence with one on other...

**Comment by vincentbernat** [comment]
Providing a list of files with `--files-from` doesn't totally solve the problem as the `snapshots` subcommand will display gigantic list of files and `forget` subcommand doesn't work as expected.

**Comment by danielsharvey** [comment]
I think @vincentbernat's [PR](https://github.com/restic/restic/pull/2311) is an effective solution to this requirement. 

_Summary: Allow gitignore-style negative patterns to specify exclude rules for both backup and restore._

I've been using it effectively for several weeks now. 

Notably, this enables me to simplify my snapshots listing, which was quite verbose previously:
```
b951f6a2  2019-06-15 11:30:18  elvandar    manual      /Users/daniel/Desktop
                                                       /Users/daniel/Documents
                                                       <lots more...>
```
to:
```
d0c0bed1  2019-06-18 08:20:57  elvandar                /Users/daniel
```

Also, I've implemented an effective (not perfect but working) solution providing [continuous backup](https://github.com/restic/restic/issues/1502) using this feature (will share details shortly).

**Comment by prosoitos** [comment]
Very interested in this feature as well.

My use case:

I have 2 restic backup repositories:

- one with high importance files that I backup daily. Those are all my personal files, the code I write, etc.
- one with low importance files that I backup weekly. That is pretty much my whole system, excluding things that it makes no sense to backup

This allows me to:

1/ have a full backup of my system
2/ run quick daily backups on what is important to me without having to run the full backup (I want to backup my files everyday, but it's ok if my system files, packages and the rest or not updated every day, but only once a week)

With an `--include` flag, it would make this a breeze. I would have:

- a file with things I never want to backup
- a file with my high importance things which I would `--include` in my daily backups and `--exclude` in my weekly ones.

As things currently are, I need 3 files and one of them needs to be updated very often.
